### PR TITLE
wrote different exceptions for different error cases

### PR DESCRIPTION
**Pull request by ljeyaratnam**
- changed the deserialser to use jackson
- added new exceptions to the current exception mapper logic
- using hibernate validator to check if fields are not null
- changed all write controller to use jackson translator

### CODE REVIEW

**Comment by MartinsIrbe** [comment]
LGTM

File: src/main/java/org/atlasapi/input/DefaultJacksonModelReader.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+                "false", Boolean.FALSE,
+                "1", Boolean.TRUE,
+                "0", Boolean.FALSE
+        );
+
+        @Override
+        public Boolean deserialize(JsonParser jsonParser,
+                DeserializationContext context)
+                throws IOException {
+            return boolMap.get(jsonParser.getText());
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/input/DefaultJacksonModelReader.java]: Do we need to support nulls or empty strings here?

File: src/main/java/org/atlasapi/input/DefaultJacksonModelReader.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+        @Override
+        public Country deserialize(JsonParser parser, DeserializationContext context)
+                throws IOException {
+
+            JsonNode node = parser.getCodec().readTree(parser);
+
+            String jsonString = node.get("code").textValue();
+
+            return Countries.fromCode(jsonString);
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/input/DefaultJacksonModelReader.java]: Ditto, do we need to support null or empty string?

File: src/main/java/org/atlasapi/query/v2/ContentWriteController.java [hunk 3]
```diff
@@ -105,22 +118,34 @@ private Void deserializeAndUpdateContent(HttpServletRequest req, HttpServletResp
             inputStreamBytes = IOUtils.toByteArray(req.getInputStream());
             InputStream inputStream = new ByteArrayInputStream(inputStreamBytes);
             inputContent = writeExecutor.parseInputStream(inputStream);
+        } catch (UnrecognizedPropertyException |
+                JsonParseException |
+                ConstraintViolationException exception) {
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/query/v2/ContentWriteController.java]: Code style. Exception variable should be named `e`

File: src/main/java/org/atlasapi/query/v2/ContentWriteController.java [hunk 4]
```diff
@@ -105,22 +118,34 @@ private Void deserializeAndUpdateContent(HttpServletRequest req, HttpServletResp
...
-            return error(resp, HttpStatusCode.FORBIDDEN.code());
-        }
 
-        if (Strings.isNullOrEmpty(content.getCanonicalUri())) {
-            return error(resp, HttpStatusCode.BAD_REQUEST.code());
+            return error(
+                    req,
+                    resp,
+                    AtlasErrorSummary.forException(new ForbiddenException(
+                            "API key does not have write permission"))
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/query/v2/ContentWriteController.java]: Code style.
```
AtlasErrorSummary.forException(new ForbiddenException(
    "API key does not have write permission"
))
```

File: src/main/java/org/atlasapi/query/v2/PeopleWriteController.java [hunk 5]
```diff
@@ -70,45 +81,64 @@ private Void deserializeAndUpdatePerson(HttpServletRequest req, HttpServletRespo
...
         }
-        
+
         if (possibleConfig.isNothing()) {
-            return error(resp, HttpStatus.UNAUTHORIZED.value());
+            return error(
+                    req,
+                    resp,
+                    AtlasErrorSummary.forException(new UnauthorizedException(
+                            "API key is unauthorised"))
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/query/v2/PeopleWriteController.java]: Code style, parenthesis

File: src/main/java/org/atlasapi/query/v2/PeopleWriteController.java [hunk 6]
```diff
@@ -70,45 +81,64 @@ private Void deserializeAndUpdatePerson(HttpServletRequest req, HttpServletRespo
...
         }
-        
+
         if (!possibleConfig.requireValue().canWrite(person.getPublisher())) {
-            return error(resp, HttpStatusCode.FORBIDDEN.code());
+            return error(
+                    req,
+                    resp,
+                    AtlasErrorSummary.forException(new ForbiddenException(
+                            "API key does not have write permission"))
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/query/v2/PeopleWriteController.java]: Ditto

File: src/main/java/org/atlasapi/query/v2/TopicWriteController.java [hunk 7]
```diff
@@ -52,74 +51,107 @@
...
         }
-        
+
         if (possibleConfig.isNothing()) {
-            return error(resp, HttpStatus.UNAUTHORIZED.value());
+            return error(
+                    req,
+                    resp,
+                    AtlasErrorSummary.forException(new UnauthorizedException(
+                            "API key is unauthorised"))
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/query/v2/TopicWriteController.java]: Ditto

File: src/main/java/org/atlasapi/query/v2/TopicWriteController.java [hunk 8]
```diff
@@ -52,74 +51,107 @@
...
         }
-        
+
         if (!possibleConfig.requireValue().canWrite(topic.getPublisher())) {
-            return error(resp, HttpStatusCode.FORBIDDEN.code());
+            return error(
+                    req,
+                    resp,
+                    AtlasErrorSummary.forException(new ForbiddenException(
+                            "API key does not have write permission"))
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/query/v2/TopicWriteController.java]: Ditto

**Comment by Dalamar42** [comment]
Aside from double-checking whether some deserialisers need to support nulls, empty strings and some code style comments LGTM

File: src/main/java/org/atlasapi/query/QueryWebModule.java [hunk 9]
```diff
@@ -377,11 +377,12 @@ ChannelWriteController channelWriteController() {
         return ChannelWriteController.create(
                 configFetcher,
                 channelStore,
-                DefaultGsonModelReader.create(),
+                new DefaultJacksonModelReader(),
```
Comment (Spartakovic) [inline src/main/java/org/atlasapi/query/QueryWebModule.java]: why?

**Pull request merged**